include/nbl/video/utilities/CArithmeticOps.h

```diff
@@ -0,0 +1,286 @@
+		}
+
+		static inline void updateDescriptorSet(ILogicalDevice* device, IGPUDescriptorSet* set, const asset::SBufferRange<IGPUBuffer>& input_range, const asset::SBufferRange<IGPUBuffer>& scratch_range)
+		{
+			IGPUDescriptorSet::SDescriptorInfo infos[2];
+			infos[0].desc = input_range.buffer;
+			infos[0].info.buffer = {input_range.offset,input_range.size};
+			infos[1].desc = scratch_range.buffer;
+			infos[1].info.buffer = {scratch_range.offset,scratch_range.size};
+
+			video::IGPUDescriptorSet::SWriteDescriptorSet writes[2];
+			for (auto i=0u; i<2u; i++)
+			{
+				writes[i] = { .dstSet = set,.binding = i,.arrayElement = 0u,.count = 1u,.info = infos+i };
+			}
+
+			device->updateDescriptorSets(2, writes, 0u, nullptr);
+		}
+
+		static inline void dispatchHelper(
+			IGPUCommandBuffer* cmdbuf, const video::IGPUPipelineLayout* pipeline_layout, const DefaultPushConstants& pushConstants, const DispatchInfo& dispatchInfo,
+			const uint32_t srcBufferBarrierCount, const IGPUCommandBuffer::SBufferMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier>* srcBufferBarriers,
+			const uint32_t dstBufferBarrierCount, const IGPUCommandBuffer::SBufferMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier>* dstBufferBarriers
+		)
+		{
+			cmdbuf->pushConstants(pipeline_layout,asset::IShader::ESS_COMPUTE,0u,sizeof(DefaultPushConstants),&pushConstants);
+			if (srcBufferBarrierCount)
+			{
+				IGPUCommandBuffer::SPipelineBarrierDependencyInfo info = { .bufBarriers = { srcBufferBarriers, 1} };
+				cmdbuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS::EDF_NONE,info);
+			}
+			cmdbuf->dispatch(dispatchInfo.wg_count,1u,1u);
+			if (srcBufferBarrierCount)
+			{
+				IGPUCommandBuffer::SPipelineBarrierDependencyInfo info = { .bufBarriers = { dstBufferBarriers, 1} };
+				cmdbuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS::EDF_NONE,info);
+			}
+		}
+
+        inline core::smart_refctd_ptr<asset::ICPUShader> createBaseShader(const char* shaderFile, const E_DATA_TYPE dataType, const E_OPERATOR op, const uint32_t scratchSz) const
+        {
+            auto system = m_device->getPhysicalDevice()->getSystem();
+            core::smart_refctd_ptr<const system::IFile> hlsl;
+            {
+                auto loadBuiltinData = [&](const std::string _path) -> core::smart_refctd_ptr<const nbl::system::IFile>
+                {
+                    nbl::system::ISystem::future_t<core::smart_refctd_ptr<nbl::system::IFile>> future;
+                    system->createFile(future, system::path(_path), core::bitflag(nbl::system::IFileBase::ECF_READ) | nbl::system::IFileBase::ECF_MAPPABLE);
+                    if (future.wait())
+                        return future.copy();
+                    return nullptr;
+                };
+
+                // hlsl = loadBuiltinData("nbl/builtin/hlsl/scan/indirect_reduce.hlsl"); ??
+                hlsl = loadBuiltinData(shaderFile);
+            }
```

Comment:
this is super convoluted, why is the user supposed to know the path to the file you'll load?